src/AppInstallerCLICore/Workflows/ManifestComparator.cpp

```diff
@@ -825,19 +849,64 @@ namespace AppInstaller::CLI::Workflow
         const Manifest::ManifestInstaller& first,
         const Manifest::ManifestInstaller& second)
     {
+        // The priority will still be used as a tie-break between weak results.
+        std::optional<std::string_view> firstWeakComparator;
+        bool firstWeakComparatorResult = false;
+
         for (auto comparator : m_comparators)
         {
-            if (comparator->IsFirstBetter(first, second))
+            details::ComparisonResult forwardCompare = comparator->IsFirstBetter(first, second);
+            details::ComparisonResult reverseCompare = comparator->IsFirstBetter(second, first);
+
+            // Should not happen, but if it does it points at a serious bug that should be fixed.
+            if (forwardCompare != details::ComparisonResult::Negative && reverseCompare != details::ComparisonResult::Negative)

```

Comment:
I had added a few other comments above, but removed them and am condensing them here for consideration.

Basically, each comparison assumes that the first option is better if the preference is equal. Take for example when `m_preference` for installer type is empty - In this case, all installer types should be considered equal. However, both a forward compare and reverse compare will return a negative comparison result since the code has a preference to the first when there is no other preference specified. 

In addition, the `negative` and `positive` structure of the comparison result seems to ignore the implied case of equality. Shouldn't there be a `ComparisonResult::Neutral` which is used in the comparator?

Reply:
> However, both a forward compare and reverse compare will return a negative comparison result since the code has a preference to the first when there is no other preference specified.

Yes, I don't see the problem.  If it is related to the line selected for this comment, it will only be true when both are positive.  Both being negative should be fine.

The function is explicitly named `IsFirstBetter` to try and remove the semantic ambiguity of `compare` or other similar words.  They also happen to align with the semantics of `operator<` if "less than" means "a better choice of installer".  The same rules apply: `a < b => true` implies `b < a => false`, but `a < b => false` does not imply anything about `b < a`.  The equivalence relationship is then based on both results being false/negative.  Introducing an equality into the comparison functions would make them more complicated than is necessary.

Yes, all of this leads to the potential for weak ordering, and thus choosing simply based on being first in the list of options.  But that will only happen in cases where there is no implied ordering, and then I think it is actually a good thing to give the manifest author control in those cases (rare as they are).

Reply:
Ah, thanks for the great explanation! I was mis-reading the case of what would happen when they were both negative.